### PR TITLE
Add new MODIFY_GRID_CONNECTIONS to voltmeter

### PR DESCRIPTION
**Pull request by Coolthulhu**
#### Summary

SUMMARY: Features "Add new MODIFY_GRID_CONNECTIONS to voltmeter"

#### Purpose of change

Being able to extend grids.

#### Describe the solution

Voltmeter gets a new function that lets it connect grids. It's very clunky, but usable.

To connect grids:
* Activate voltmeter
* Pick the use function
* Pick the direction you want to connect current map square to
* Pay the cost multiplied by the number of map squares in the resulting grid, or 0 if it does not extend current grid

The function is a "toggle" and will say if you are going to connect or disconnect the square.
Disconnecting is free.

Cost is a requirement, not a recipe.

Functionality that didn't make it into this PR:
* Requirements for the grid, like "it has at least one wall tile" - you can create a free-floating grid by connecting tiles while on a helicopter, if you manage to keep it in the air
* UI that would tell you what size of grid are you connecting to before you pick the direction
* UI for grid display - right now, you have to turn graphical overmap off and use the debug function to show grids, or figure it out semi-blindly by groping around with the voltmeter
* (Not sure if desirable) Cost of connecting that doesn't spoil grid size, allows large grids
* Long action for connecting the grids - right now, it's instant
* Reclaiming anything when disconnecting grids

#### Describe alternatives you've considered

* Spending more time on the UI
* Delaying it until I can figure out how to do it better
* Not expanding grids

#### Testing

* Spawn voltmeter
* Connect two "empty" grids together
* Verify, using debug "show grids", that they are both in the same grid now
* Verify that solar panel in one will charge a battery in the other one
* Try to connect two house grids
* Verify that the cost is proportional to number of tiles in the resulting grid (typically 4, for two floors and two roofs)
* Connect and verify that they are connected
* Disconnect them and verify that they are disconnected

### CODE REVIEW

**Comment by joveeater** [comment]
LGTM.

I've tested this and done a code review. Obviously there's the bit that msvc is complaining about, apparently it doesn't like std::distance there for some reason but other than that, it all looks good.

File: src/overmapbuffer.cpp [hunk 1]
```diff
@@ -1722,3 +1723,88 @@ overmapbuffer::electric_grid_connectivity_at( const tripoint_abs_omt &p )
...
+
+    overmap_with_local_coords lhs_omc = get_om_global( lhs );
+    overmap_with_local_coords rhs_omc = get_om_global( rhs );
+
+    const tripoint *lhs_iter = std::find( six_cardinal_directions.begin(),
+                                          six_cardinal_directions.end(),
+                                          tripoint( coord_diff.x(), coord_diff.y(), coord_diff.z() ) );
+    const tripoint *rhs_iter = std::find( six_cardinal_directions.begin(),
+                                          six_cardinal_directions.end(),
+                                          -tripoint( coord_diff.x(), coord_diff.y(), coord_diff.z() ) );
```
Comment (olanti-p) [inline src/overmapbuffer.cpp:1748]:
```suggestion
    const auto lhs_iter = std::find( six_cardinal_directions.begin(),
                                     six_cardinal_directions.end(),
                                     coord_diff.raw() );
    const auto rhs_iter = std::find( six_cardinal_directions.begin(),
                                     six_cardinal_directions.end(),
                                     -coord_diff.raw() );
```

File: src/overmapbuffer.cpp [hunk 2]
```diff
@@ -1722,3 +1723,88 @@ overmapbuffer::electric_grid_connectivity_at( const tripoint_abs_omt &p )
...
```
Comment (olanti-p) [inline src/overmapbuffer.cpp:1789]:
```suggestion
    const auto lhs_iter = std::find( six_cardinal_directions.begin(),
                                     six_cardinal_directions.end(),
                                     coord_diff.raw() );
    const auto rhs_iter = std::find( six_cardinal_directions.begin(),
                                     six_cardinal_directions.end(),
                                     -coord_diff.raw() );
```

**Review by olanti-p — COMMENTED** [review]
> there's the bit that msvc is complaining about, apparently it doesn't like std::distance there for some reason

It doesn't like implicit casts between an iterator and a raw pointer, this ought to fix it.
Also nitpick: there's `.raw()` method to convert from typed `tripoint` to untyped.

File: src/overmapbuffer.cpp [hunk 3]
```diff
@@ -1722,3 +1723,88 @@ overmapbuffer::electric_grid_connectivity_at( const tripoint_abs_omt &p )
 
     return ret;
 }
+
+bool overmapbuffer::add_grid_connection( const tripoint_abs_omt &lhs, const tripoint_abs_omt &rhs )
+{
+    if( project_to<coords::om>( lhs ) != project_to<coords::om>( rhs ) ) {
```
Comment (olanti-p) [inline src/overmapbuffer.cpp:1729]: It doesn't let me connect omts on different z-levels (e.g. house room and house roof) - is this intentional?
If not, this fixes it.
```suggestion
    if( project_to<coords::om>( lhs ).xy() != project_to<coords::om>( rhs ).xy() ) {
```

**Pull request merged**